feat: Add select / deselect all to table filter menu #602

aalencar opened the pull request:
Add Select and Deselect all functionality to filter menu

![2022-01-14 14 12 03](https://user-images.githubusercontent.com/15820796/149520767-0d7f9209-fa4c-41a2-b3b6-1a57a2206b53.gif)

mturoci commented:
Thanks @aalencar. Please have a look at `feat/alternative-issue-602` branch for an alternative, non-hacky, solution and let me know what you think

aalencar commented:
Thanks for providing an alternative, @mturoci !

I like the idea of having a new component for Contextual Menu and not using `setTimeout` to trigger the changes.

But I noticed that the current solution is broken when we have 2 filterable columns. The bug is in `onFilterChange`. Working on a fix.

![2022-01-17 17 29 39](https://user-images.githubusercontent.com/15820796/149807058-449c93ee-eeb3-4434-9e37-b9d3fc3870e2.gif)

aalencar commented:
@mturoci I fixed the issue above, and added other changes to `feat/alternative-issue-602`.

One of the missing fixes was persisting the checkbox status after the contextual menu was closed. For that we needed to set the `checked` property of each item in the `filters` state upon initialisation. This raises a problem: by setting `checked` we now have a controlled component and we need to set ContextualMenu state when the checkboxes are clicked. To achieve that I passed a callback to `onFilterChange` that has the access to ContextualMenu `filters` used to manage the checkbox state.

Let me know what you think of the solution.

aalencar commented:
I just realized that I might have overlooked the logic for tags. Taking a look at it.

mturoci commented:
@aalencar if you decided to take the approach proposed in `feat/alternative-issue-602`, can you please also switch branch for this PR? The review will be easier that way.

aalencar commented:
@mturoci done.

mturoci commented:
> To achieve that I passed a callback to onFilterChange that has the access to ContextualMenu filters used to manage the checkbox state.

This seems to be way more complex than necessary. IIUC, what you do basically is:

* checkbox is clicked
* call parent `onFilterChange`
* from within top-level component (`onFilterChange`) call the callback to update context menu state

I would suggest a simpler approach, which doesn't require any callback:

* checkbox is clicked
* update menu state right away, then call original `onFilterChange` callback

Or maybe I am missing something?

aalencar commented:
> update menu state right away, then call original onFilterChange callback

`onFilterChange()` is not the event handler but instead creates a custom handler for each filter.

We can't update right away because we have to update the state from within the handler which is returned by `onFilterChange`, and that's why we also pass a callback to `onFilterChange`.

mturoci commented:
> We can't update right away because we have to update the state from within the handler which is returned by onFilterChange, and that's why we also pass a callback to onFilterChange.

I still don't understand, why would we have to update the state within the handler and not directly? You can refactor the code, no need to stick with what is there currently.

aalencar commented:
> why would we have to update the state within the handler and not directly

Because we want to update the ContextualMenu's state on click and that has to be done from within the handler.

aalencar commented:
@mturoci applied your suggestions. Ready to be reviewed again.

mturoci commented on ui/src/table.tsx:
```diff
@@ -119,7 +119,7 @@ type WaveColumn = Fluent.IColumn & {
 type DataTable = {
   model: Table
-  onFilterChange: (filterKey: S, filterVal: S) => (e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => void
+  onFilterChange: (filterKey: S, filterVal: S, checked?: boolean) => void
```
nit
```suggestion
  onFilterChange: (filterKey: S, filterVal: S, checked?: B) => void
```

mturoci commented on ui/src/table.tsx:
```diff
@@ -208,7 +217,51 @@ const
+    const
+      isFilterChecked = (data: S, key: S) => !!selectedFiltersRef.current && selectedFiltersRef.current[data]?.includes(key),
+      [filters, setFilters] = React.useState(col.cellType?.tag
+        ? Array.from(listProps.items.reduce((_filters, { key, text, data }) => {
+          key.split(',').forEach(key => _filters.set(key, { key, text, data }))
+          return _filters
```
Didn't check, but it seems like it's missing handling `checked`.

aalencar replied:
Good catch.

mturoci commented on ui/src/table.tsx:
```diff
@@ -208,7 +217,51 @@ const
+        setFilters(filters.map(i => ({ ...i, checked: false })))
+        setFiltersInBulk(col.key, [])
+      },
+      getOnFilterChangeHandler = (data: S, key: S) => (_ev?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: boolean) => {
```
Odd naming, handlers usually start with on rather than get.

nit
```suggestion
      getOnFilterChangeHandler = (data: S, key: S) => (_ev?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
```

aalencar replied:
Not a handler, it's a high-order function that creates a handler, hence the name

mturoci replied:
Agree, fair point.

mturoci commented on ui/src/table.tsx:
```diff
@@ -543,26 +579,25 @@ export const
         initGroups()
       },
-      onFilterChange = React.useCallback((filterKey: S, filterVal: S) => (_e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
+      onFilterChange = React.useCallback((filterKey: S, filterVal: S, checked) => {
```
Missing type for checked

mturoci commented on ui/src/table.tsx:
```diff
@@ -208,7 +217,51 @@ const
+  ContextualMenu = ({ onFilterChange, col, listProps, selectedFiltersRef, setFiltersInBulk }: ContextualMenuState) => {
+    const
+      isFilterChecked = (data: S, key: S) => !!selectedFiltersRef.current && selectedFiltersRef.current[data]?.includes(key),
+      [filters, setFilters] = React.useState(col.cellType?.tag
```
Would go for `menuFilters` or something like that so that reader is sure which ones are meant.

aalencar replied:
Indeed, surely more meaningful.

mturoci reviewed (changes requested):
Thanks @aalencar. A few minor comments. I didn't try your solution yet, will do tmr.

mturoci commented on ui/src/table.tsx:
```diff
@@ -543,26 +579,25 @@ export const
         setGroupByKey(option.key as S)
         initGroups()
       },
-      onFilterChange = React.useCallback((filterKey: S, filterVal: S) => (_e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
+      onFilterChange = React.useCallback((filterKey: S, filterVal: S, checked?: B) => {
         setSelectedFilters(selectedFilters => {
-          const filters = selectedFilters || {}
-
+          let filters: Dict<S[]> | null = null
+          filters = selectedFilters || {}
```
What's the reason for this change?

aalencar replied:
It was probably a mistake during the refactoring. Reverted.

mturoci reviewed (approved):
LGTM @aalencar! Just one minor comment from me. Also, let's merge this after `0.20` is released.
